src/devices/gen.jl

```diff
@@ -0,0 +1,210 @@
+
+function add_pm_cost!(
+    PM_gen::Dict{String, Any},
+    op_cost::T,
+    base::Float64,
+) where {T <: PSY.ThreePartCost}
+    PM_gen["startup"] = PSY.get_start_up(op_cost)
+    PM_gen["shutdown"] = PSY.get_shut_down(op_cost)
+    add_pm_var_cost!(PM_gen, PSY.get_variable(op_cost), base, PSY.get_fixed(op_cost))
+    return PM_gen
+end
+
+function add_pm_cost!(
+    PM_gen::Dict{String, Any},
+    op_cost::T,
+    base::Float64,
+) where {T <: PSY.TwoPartCost}
+    PM_gen["shutdown"] = 0.0
+    PM_gen["startup"] = 0.0
+    add_pm_var_cost!(PM_gen, PSY.get_variable(op_cost), base, PSY.get_fixed(op_cost))
+    return PM_gen
+end
+
+# polynomial cost
+function add_pm_var_cost!(
+    PM_gen::Dict{String, Any},
+    var_cost::PSY.VariableCost{Tuple{Float64, Float64}},
+    base::Float64,
+    fixed::Float64,
+)
+    cost = PSY.get_cost(var_cost)
+    PM_gen["model"] = 2
+    PM_gen["cost"] = Vector{Float64}()
+    PM_gen["ncost"] = cost[1] != 0.0 ? 3 : cost[2] != 0.0 ? 2 : 0
+    PM_gen["ncost"] == 0 && return PM_gen
+
+    for idx in (4 - PM_gen["ncost"]):2
+        push!(PM_gen["cost"], cost[idx])
+        PM_gen["cost"] = PM_gen["cost"] .* base
+    end
+    push!(PM_gen["cost"], fixed)
+    return PM_gen
+end
+
+# pwl cost
+function add_pm_var_cost!(
+    PM_gen::Dict{String, Any},
+    var_cost::PSY.VariableCost{Vector{Tuple{Float64, Float64}}},
+    base::Float64,
+    fixed::Float64,
+)
+    PM_gen["model"] = 1
+    PM_gen["ncost"] = length(var_cost)
+    PM_gen["cost"] = Vector{Float64}()
+    for c in PSY.get_cost(var_cost)
+        push!(PM_gen["cost"], last(c))
+        push!(PM_gen["cost"], first(c) + fixed)
+    end
+    return PM_gen
+end
+
+# scalar cost
+function add_pm_var_cost!(
+    PM_gen::Dict{String, Any},
+    var_cost::PSY.VariableCost{Float64},
+    base::Float64,
+    fixed::Float64,
+)
+    PM_gen["model"] = 2
+    PM_gen["ncost"] = length(var_cost)
+    PM_gen["cost"] = [PSY.get_cost(var_cost) + fixed]
+    return PM_gen
+end
+
+function pm_gen_core(gen::T, ix::Int) where {T <: PSY.Generator}
```

Comment:
why `core`?

Reply:
because it's common to all `<:Generator`s

Reply:
lol ok, it isn't obvious. we usually use the word common in the other packages.